---
title: "EMOGEA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EMOGEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



#1 Usage

```{r echo = T, eval = F}
#-----------------------
# (1) Overview
#----------------------
# Error Modeled Gene Expression Analysis (EMOGEA), an R package for the analysis of RNA-Seq gene expression data. EMOGEA incorporates measurement uncertainties in
# the analysis of differential expression and is specifically suited for transcriptomics studies in which low-count transcripts with small fold-changes lead to significant
# biological effects. Such transcripts include signaling mRNAs and non-coding RNAs (ncRNA) which are known to exhibit low levels of expression. The package handles missing 
# values by associating disproportionately large uncertainties to those measurements, making it particularly useful for single-cell RNA-Seq measurements. It is specifically 
# suited for ordinal data as it implements a constrained alternating least-squares (ALS) approach that allows waves of expression profiles to be visualized against the ordinal 
# variable. For differential expression analysis, EMOGEA has a much higher true positivity rate (TPR) and a vanishingly small false negativity rate (FNR) compared to common 
# approaches.  
#-------------------------
# (2) Setting up the data
# ------------------------
# We start with a count matrix where each row is a gene and each column is a sample (for bulk RNAseq) or a single cell (for scRNAseq). 
# These can be obtained by mapping read sequences to a reference genome and then counting the number of reads mapped to the exons of each gene. 
# (See, for example, the Rsubread package which does both tasks.) Alternatively, pseudo-alignment methods can be used to quantify the abundance 
# of each transcript in each sample.  Alternatively, we can also accept data arising from Affymetrix arrays. In this example we will use yeast data 
# generated by  Gierli≈Ñski et al., [PMID: 26206307].

library(EMOGEA)
data(yeastExample)

#---------- (i) Get variable ------------------
#
expressionData <- yeastExample$expressionData
metaData <- yeastExample$metaData
sampleColumn <- yeastExample$sampleColumn
conditionColumn <- yeastExample$conditionColumn

#---------- (ii) Prepare data -------------------
#
prepareDataOutput <- prepareData(
  expressionData = expressionData,
  metaData = metaData,
  sampleColumn = sampleColumn,
  conditionColumn = conditionColumn,
  applyLogTransformation = FALSE)

# (iii) Maximum Likelihood Projection
projectionOutput <- mlProjection(
  expressionMatrix = prepareDataOutput$expressionMatrix,
  errorCovarianceMatrix = prepareDataOutput$errorCovarianceMatrix)

# Curve resolution
curveResOutput <- multivariateCurveResolution(
  expressionMatrix = prepareDataOutput$expressionMatrix,
  residualMatrix = prepareDataOutput$residualMatrix)

# Curve resolution (without residual matrix)
curveResOutput <- multivariateCurveResolution(expressionMatrix = prepareDataOutput$expressionMatrix)
```










